test/binder-call/binder-call.c

```diff
@@ -0,0 +1,740 @@
+                            }
+                            printf("] ");
+                        } else {
+                            guint i;
+                            gsize count;
+                            GBinderHidlVec* vec = (GBinderHidlVec*)(((char*)buf)+offset);
+                            int off;
+
+                            count = vec->count;
+
+                            printf("[ ");
+                            off = 0;
+                            for (i = 0; i < count; i++) {
+                                off += go_through_reply_ast(app, NULL, t->data, vec->data.ptr + off, cur_pass);
+                            }
+                            printf("] ");
+                        }
+                    }
+                    offset += sizeof(GBinderHidlVec);
+                } else {
+                    guint i;
+                    gint32 vl;
+                    gbinder_reader_read_int32(&app->reader, &vl);
+                    printf("[ ");
+                    for (i = 0; i < vl; i++) {
+                        go_through_reply_ast(app, NULL, t->data, NULL, cur_pass);
+                    }
+                    printf("] ");
+                }
+                if (cur_pass == PRINT_REPLY) GDEBUG("vectorend");
+                break;
+            }
+            default:
+                GERR("unknown type: %d\n", t->type);
+                break;
+        };
+
+        if (tt) break;
+    }
+
+    return offset;
+}
+
+void
+go_through_ast(
+    App* app,
+    struct transaction_and_reply* ast,
+    gboolean transaction)
+{
+    if (ast) {
+        if (transaction && ast->tree_transaction) {
+            go_through_transaction_ast(app, ast->tree_transaction, -1, NULL, BUILD_TRANSACTION, 0);
+        } else if (!transaction && ast->tree_reply) {
+            GDEBUG("REPLY:");
+            go_through_reply_ast(app, ast->tree_reply, NULL, NULL, PRINT_REPLY);
+            printf("\n");
+        }
+    }
+}
+
+void
+free_ast_transaction_tree(
+    gpointer data)
+{
+    struct value_info* v = data;
+
+    if (v->type == STRUCT_TYPE || v->type == VECTOR_TYPE) {
+        g_list_free_full(v->value, free_ast_transaction_tree);
+    } else {
+        g_free(v->value);
+    }
+
+    g_free(v);
+}
+
+void
+free_ast_reply_tree(
+    gpointer data)
+{
+    struct type_info* t = data;
+
+    if (t->type == VECTOR_TYPE) {
+        free_ast_reply_tree(t->data);
+    } else if (t->type == STRUCT_TYPE) {
+        g_list_free_full(t->data, free_ast_reply_tree);
+    }
+
+    g_free(t);
+}
+
+void
+free_ast(
+    struct transaction_and_reply* ast)
+{
+    if (ast) {
+        g_list_free_full(ast->tree_transaction, free_ast_transaction_tree);
+        g_list_free_full(ast->tree_reply, free_ast_reply_tree);
+        g_free(ast);
+    }
+}
+
+static
+void
+app_run(
+    App* app)
+{
+    const AppOptions* opt = app->opt;
+    char* iface = opt->iface ? g_strdup(opt->iface) : NULL;
+
```

Comment:
Empty lines randomly appearing here and there give the impression of randomly generated code.